### ISSUE TITLE
CT: countProperty should match array length

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/natural-selection#89]
This problem started at 12:17am this morning, and occurred again at 11:16am.  I can reproduce locally in the State Wrapper with `?sim=natural-selection&phetioDebug&fuzz`.

```
natural-selection : phet-io-state-fuzz : unbuilt
https://bayes.colorado.edu/continuous-testing/ct-snapshots/1590599772238/phet-io-wrappers/state/?sim=natural-selection&phetioDebug&fuzz&wrapperContinuousTest=%7B%22test%22%3A%5B%22natural-selection%22%2C%22phet-io-state-fuzz%22%2C%22unbuilt%22%5D%2C%22snapshotName%22%3A%22snapshot-1590599772238%22%2C%22timestamp%22%3A1590603649651%7D
Uncaught Error: Assertion failed: countProperty should match array length.
Error: Assertion failed: countProperty should match array length.
at window.assertions.assertFunction (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1590599772238/assert/js/assert.js:22:13)
at https://bayes.colorado.edu/continuous-testing/ct-snapshots/1590599772238/tandem/js/PhetioGroup.js:69:9
at TinyEmitter.emit (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1590599772238/axon/js/TinyEmitter.js:69:53)
at https://bayes.colorado.edu/continuous-testing/ct-snapshots/1590599772238/axon/js/Emitter.js:33:29
at Emitter.execute (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1590599772238/axon/js/Action.js:225:18)
at Emitter.emit (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1590599772238/axon/js/Emitter.js:58:19)
at PhetioStateEngine.setState (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1590599772238/phet-io/js/PhetioStateEngine.js:219:26)
at PhetioEngineIO.implementation (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1590599772238/phet-io/js/PhetioEngineIO.js:221:43)
at PhetioCommandProcessor.processCommand (https://bayes.colorado.edu/continuous-testing/ct-snapshots/1590599772238/phet-io/js/phetioCommandProcessor.js:267:51)
at https://bayes.colorado.edu/continuous-testing/ct-snapshots/1590599772238/phet-io/js/phetioCommandProcessor.js:160:36
id: Bayes Chrome
Snapshot from 5/27/2020, 11:16:12 AM
```

**Comment by zepumph** [comment]
This issue takes the same form as https://github.com/phetsims/john-travoltage/issues/348 and https://github.com/phetsims/projectile-motion/issues/229. I will take a look!

**Comment by pixelzoom** [comment]
Since the code in WolfCollection is temporary, I put a `if ( !phet.joist.sim.isSettingPhetioStateProperty.value )` around it.

But WolfCollection was added within the past 2 hours, and this problem started at 12:17am this morning.  And CT was clear for 4 cycles after my last commits yesterday.   So I'm confused about why this suddenly started at 12:17am.

**Comment by zepumph** [comment]
Does it have to do with https://github.com/phetsims/natural-selection/commit/6754aa8168bee7c191418994ea12e9a1f2b19c13? The timing on that seems likely.

**Comment by pixelzoom** [comment]
I applied the patch. After 10 minutes of testing locally with `?sim=natural-selection&phetioDebug&fuzz`, no problem.  I reverted the patch, tested for another 5 minutes, still no problem.  I'm monitor CT and close this issue if the error does not reoccur.

**Comment by pixelzoom** [comment]
@zepumph I believe you're confusing AM and PM. https://github.com/phetsims/natural-selection/commit/6754aa8168bee7c191418994ea12e9a1f2b19c13 was at 12:08pm.   The first CT failure was at 12:17am, almost 12 hours before that commit.

**Comment by zepumph** [comment]
Hmm, not on my system, maybe it pushed very delayed? I see it as this morning in multiple sources. The time stamp in https://github.com/phetsims/natural-selection/commit/6754aa8168bee7c191418994ea12e9a1f2b19c13:

![image](https://user-images.githubusercontent.com/6856943/83076378-e7b53980-a021-11ea-8ab8-1622bbaf15c0.png)


As well as in my webstorm log and git log:

```
809b0358cb2dbdd63bbfe11f0e868306416ec0a9
  Wed May 27 14:03:52 2020 -0600 cmalley@pixelzoom.com make wolves move around


16be713a74d8752eaec14ba5dd303f766ab538fc
  Wed May 27 13:11:52 2020 -0600 cmalley@pixelzoom.com https://github.com/phetsims/natural-selection/issues/89


930a4ababf7187d521980c4abe1e62293094d1ac
  Wed May 27 12:48:18 2020 -0600 cmalley@pixelzoom.com Wolf.move


6754aa8168bee7c191418994ea12e9a1f2b19c13
  Wed May 27 12:08:12 2020 -0600 cmalley@pixelzoom.com first pass at creating wolves dynamically using PhetioGroup


b4613d7b6af7eed44d4ec959666d29223057eb50
  Wed May 27 10:21:11 2020 -0600 cmalley@pixelzoom.com compute number of wolves

```

**Comment by pixelzoom** [comment]
@zepumph I'm not sure how to interpret the previous comment.  Do you not see the CT error at 12:17 AM MDT, and my first today at 10:12 AM MDT?     The CT error predates any of today's commits by almost 10 hours.

**Comment by pixelzoom** [comment]
This error is still occurring in CT as of 8:20am this morning. It's not occurring on every cycle.

**Comment by pixelzoom** [comment]
@zepumph and I discussed this on Zoom.  He identified that problem as occurring when wolves kill bunnies.  That occurs in WolfCollection.apply, which was being called from a listener for a Property that was being restored.   So the fix (in the above commit) was once again to bracket the code with `if ( !phet.joist.sim.isSettingStateProperty.value )`.  

Since this appears to be a common problem, @zepumph is going to commit error checking something like the patch in https://github.com/phetsims/natural-selection/issues/89#issuecomment-634875050, so that this category of problem is easier to identify and diagnose.

**Comment by zepumph** [comment]
A few more specifics from diving into this further. I was getting confused as to why this was only being triggered from BunnyNodeGroup and not from BunnyGroup. 

I also found this very reproducible by doing the following:
* Load state wrapper with no special query parameters
* Add a mate
* add wolves
* wait for wolves to eat a bunny
*  The error will always occur here. It will show that bunnyNodeGroup.countProperty wanted to be 1, but was actually 0.

This confused me for some time, but it seems to be that the state engine only creates one bunnyNode (as it is supposed to), and then `WolfCollection.apply()` called during the listener will proceed in killing the other Bunny. Two important points here
1. The listeners was causing double work being done. This aligns with the main bug that the countProperty assertion is trying to fix. The state engine should not trigger sim logic that then alters the state of the dynamic, stateful, PhetioGroups.
2. Natural Selection follows the "only the models are stateful" pattern, in which the view listens to the model's dynamic element creation/disposal and then creates/disposes the Node accordingly. The special piece that NS adds is that a Bunny dying causes the  Node to be disposed, but not the Bunny itself. The assertions that I added above would catch this bugginess for statful groups (like BunnyGroup), but cannot easily do that for the corresponding Node group (BunnyNodeGroup). Thus @pixelzoom, we will still get `countProperty should match array length` assertions in this case, and not the more specific error. I can't think of a patch that will land us right in the proper stack trace where mischief is occurring. 

The above commits add assertions to try to eagerly catch dynamic elements created or destroyed during state setting when not by the state engine itself. It added a bit of boiler-plate. More than I would normally desire to just add an assertion, but @pixelzoom and I left our call with the sentiment that this type of assertion is likely going to occur often, and it isn't enough to just have a patch that we pass along from issue to issue to use when solving this. I'm open to reviewing that code as we move forward and find (or don't find) new bugs of this kind. 

I confirmed that the above fixed the bug. As a reminder, this issue had two bugs in it that triggered the same assertion. 

CT was clear the most recent round so I'm going to close this issue. Thanks @pixelzoom for working with me here!
